### ISSUE TITLE
Improve `--help` docs to denote which commands/flags apply to Formula, which to Casks, and which to both

### ISSUE THREAD

**Issue by 0xdevalias** [Homebrew/brew#16334]
### Verification

* [x]  This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at [Homebrew/homebrew-core/issues/new/choose](https://github.com/Homebrew/homebrew-core/issues/new/choose) instead.

#### Overview

The purpose of this issue is to propose an enhancement to the Homebrew `--help` documentation. The goal is to make it more user-friendly and informative, specifically by clearly delineating which commands and flags are associated with formulas, which are with casks, and which are universal. This improvement is intended to streamline the user experience, reduce confusion, and enhance the overall usability of Homebrew.

#### Outcomes

- **Clarity in Documentation:** Revise the `--help` documentation to include clear distinctions between formula-specific, cask-specific, and universal commands/flags.

#### Proposed Steps

1. **Audit Existing Documentation:** Review the current `--help` documentation to identify areas lacking clarity.
2. **Documentation Update:** Implement changes to the `--help` documentation, incorporating clear labels or sections for formulas, casks, and universal commands/flags.

##### Examples, Existing Patterns, and Potential Improvements

In `brew install` (and related commands), there is a `--build-from-source` for formula, which doesn't make sense in the context of casks. While I can infer that, without looking at the implementation specifics, it's only a guess. While in this case it's probably a fairly safe/easy guess, in other cases previously it's been less easy to distinguish.

Looking at the code for `brew install`, I can see that `--build-from-source` is after the cask section, and so looks like it only applies to formula:

_(note that the cask related code is above this)_

https://github.com/Homebrew/brew/blob/2e9d7fbcf81134fbe594d5a18b5387dad76383a1/Library/Homebrew/cmd/install.rb#L264-L273

Since there are flags for `--formula` and `--cask`, we could potentially leverage an existing pattern of using `conflicts` when `--cask`, similar to:

https://github.com/Homebrew/brew/blob/2e9d7fbcf81134fbe594d5a18b5387dad76383a1/Library/Homebrew/cmd/install.rb#L158-L160

That would cover an explicit case of `--cask`, but not an indirect case of providing only the name of a cask(s) to `brew install`.

Formulae and casks are partioned, and the cask section of the code seems to check for presence of them with `casks.any?`:

https://github.com/Homebrew/brew/blob/2e9d7fbcf81134fbe594d5a18b5387dad76383a1/Library/Homebrew/cmd/install.rb#L194-L207

Perhaps another little check could be added before here that works similar to the `conflicts` DSL, but handling the case of an implied cask. Pseudocode'ish example:

```ruby
if !formulae.any? && (args.build_bottle? || args.build_from_source?)
  // TODO: throw error/tell user/similar
```

---

An existing pattern that helped improve clarity for me was in `brew audit --help`, which explicitly states that `--new` implies `--strict --online`:

https://github.com/Homebrew/brew/blob/2e9d7fbcf81134fbe594d5a18b5387dad76383a1/Library/Homebrew/dev-cmd/audit.rb#L47-L50

And actually seems to do so in the code:

https://github.com/Homebrew/brew/blob/2e9d7fbcf81134fbe594d5a18b5387dad76383a1/Library/Homebrew/dev-cmd/audit.rb#L121-L122

I'm not super familiar with the CLI DSL, but if this were to become a more widespread pattern, I wonder if the DSL could be enhanced to make it easier to only have to define the 'implies' in one place, rather than having to separately update help text + actual code implementation. (though also might be overoptimising to do so; just pondering about how to lessen the risk of the text and 'implies' code falling out of sync; but maybe code review is good enough for that)

---

Since it seems `brew audit` will also run style checks, it might be useful to also mention that somewhere explicitly in it's help text; so that users know they don't need to run `brew style --fix foo` immediately followed by `brew audit --new foo`; when they could have just run `brew audit --fix --new foo` or similar (maybe this is something that can live in the larger documentation rather than the commands `--help` text; eg. https://github.com/Homebrew/brew/pull/16327 or similar; see specifically: https://github.com/Homebrew/brew/pull/16327#discussion_r1425267157)

https://github.com/Homebrew/brew/blob/2e9d7fbcf81134fbe594d5a18b5387dad76383a1/Library/Homebrew/dev-cmd/audit.rb#L199-L200

---

There are no doubt more examples/patterns to be found, but this seems like enough to demonstrate the idea for now.

#### Implementation Ideas

Based on the above exploration of existing patterns, there are 2 higher level ways I think this could be implemented:

- Individually in the help text of each arg/flag (eg. like the `brew audit --new` example of saying it implies `--strict --online`; but saying something like "cask only" or "formula only")
  - This could end up being fairly verbose though, and might clutter the help text a bit
- By partitioning the help text of the args/flags into 'universal' (applies to both formula and casks), 'formula', and 'casks'.
  - This feels like the cleaner way to me, making the distinction clear; but it might take some extra work in refactoring the layout, and i'm not sure if the current CLI DSL supports this sort of 'sections' for help text

### What is the motivation for the feature?

> It's also somewhat confusing sometimes to know what flags are meant only for formula, which are for casks, and which apply to all. It would be nice if the help docs reflected that (though that's OOS of this issue)
> 
> _Originally posted by @0xdevalias in https://github.com/Homebrew/brew/issues/16325#issuecomment-1854784140_

### How will the feature be relevant to at least 90% of Homebrew users?

This should make it easier for any user creating a new Formula/Cask, or using other CLI tools, to be able to understand how to use them quicker/easier, reducing friction/required background knowledge to be productive.

### What alternatives to the feature have been considered?

- Leaving things as they are.

**Comment by apainintheneck** [comment]
I agree that it would nice if users could easily see at a glance if a command is formula or cask specific. Sometimes this is spelled out explicitly in the description itself.

```
  -s, --build-from-source          Compile formula from source even if a
                                   bottle is provided. Dependencies will still
                                   be installed from bottles if they are
                                   available.
```

Other times it requires a little bit of knowledge about where certain files are placed in the operating system. The following command only works for casks.

```
      --keyboard-layoutdir         Target location for Keyboard Layouts
                                   (default: /Library/Keyboard Layouts).
```

We autogenerate the man/help pages for each command based on the arg parser which should have enough information to decide if a command is cask or formula only so it seems like we could probably automate any format that we decide upon (see `brew generate-man-completions`).

> - Individually in the help text of each arg/flag (eg. like the brew audit --new example of saying it implies --strict --online; but saying something like "cask only" or "formula only")
>    - This could end up being fairly verbose though, and might clutter the help text a bit
> - By partitioning the help text of the args/flags into 'universal' (applies to both formula and casks), 'formula', and 'casks'.
>    - This feels like the cleaner way to me, making the distinction clear; but it might take some extra work in refactoring the layout, and i'm not sure if the current CLI DSL supports this sort of 'sections' for help text

I agree that this sounds ideal but there are some formatting limitations. We try to keep the text width to a maximum of 80 characters so things stay readable when using the default terminal window size. Avoiding nesting is crucial for that reason. Sub-headings could work though or adding more context to the description as you mentioned.

**Comment by apainintheneck** [comment]
> Since it seems brew audit will also run style checks, it might be useful to also mention that somewhere explicitly in it's help text; so that users know they don't need to run brew style --fix foo immediately followed by brew audit --new foo

This is implied in two parts of the help page. The messaging could maybe be improved though.

```md
Check formula for Homebrew coding style violations. This should be run before
submitting a new formula or cask. If no formula|cask are provided, check all
locally available formulae and casks and skip style checks. Will exit with a
non-zero status if any errors are found.
```

```
      --skip-style                 Skip running non-RuboCop style checks. Useful
                                   if you plan on running brew style
                                   separately. Enabled by default unless a
                                   formula is specified by name.
```

**Comment by MikeMcQuaid** [comment]
Hey @0xdevalias, thanks for the issue but can you make sure you fill out the issue template when creating new issues in future? Thanks!

---

Onto the text:

> I wonder if the DSL could be enhanced to make it easier to only have to define the 'implies' in one place

This seems like a good idea, agreed, given how many times this is done. This is partly implemented by the fact that there are `depends_on` and `conflict` DSLs for options but these don't translate into differing `--help` text (and probably should).

> Individually in the help text of each arg/flag (eg. like the `brew audit --new` example of saying it implies `--strict --online`; but saying something like "cask only" or "formula only")
> * This could end up being fairly verbose though, and might clutter the help text a bit

This is my preferred approach although, as @apainintheneck said, it should be autogenerated.

> * By partitioning the help text of the args/flags into 'universal' (applies to both formula and casks), 'formula', and 'casks'.
>   
>   * This feels like the cleaner way to me, making the distinction clear; but it might take some extra work in refactoring the layout, and i'm not sure if the current CLI DSL supports this sort of 'sections' for help text

Less keen on the partitioning. For something like the `--help` documentation where people are often searching/`grep`ing for particular flags being DRY is actively unhelpful. Let's be repetitive here for now; it's easier to autogenerate and clearer.

**Comment by 0xdevalias** [comment]
> This is implied in two parts of the help page. The messaging could maybe be improved though.

@apainintheneck I'll admit that I've probably never fully read/paid attention to most of the help text for `brew style -h` and `brew audit -h`, as I am usually just blindly copy pasting the commands from the docs, the PR template, or [my own little helper aliases/info shortlist](https://github.com/0xdevalias/dotfiles/blob/devalias/homebrew/aliases.zsh) (that I created when it seemed too hard to figure out and remember all of the required patterns/things within Homebrew)

---

> thanks for the issue but can you make sure you fill out the issue template when creating new issues in future?

@MikeMcQuaid I think I created this from the ['reference in new issue'](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue#creating-an-issue-from-a-comment) button, which apparently bypasses the options / templates from the [new issue chooser](https://github.com/Homebrew/brew/issues/new/choose):

![image](https://github.com/Homebrew/brew/assets/753891/31f2babc-1d3f-413a-8186-403f67734d23)

I think I was trying to capture it while it was still in my mind (while working on unrelated stuff), and didn't want to forget it in the mental chaos. Will try and remember to do it the long way in future 👌🏻 

---

> This is partly implemented by the fact that there are `depends_on` and `conflict` DSLs for options but these don't translate into differing `--help` text (and probably should).

@MikeMcQuaid Yeah, I was thinking about that in relation to this issue. If the DSL's 'constraints' can automagically translate into some 'well known' additions to the help text, that would make it easier to maintain/upgrade + keep in sync. **tl;dr:** 👌🏻 

---

> Less keen on the partitioning. For something like the --help documentation where people are often searching/greping for particular flags being DRY is actively unhelpful. Let's be repetitive here for now; it's easier to autogenerate and clearer.

@MikeMcQuaid Yeah, good point about searching/grepping through the help text. Given that, I agree that autogenerated help text additions to each `--flag`/etc based on 'constraints' sounds like the better approach 👌🏻 

Given that, it's probably worth figuring out all of the 'constraints' that are possible (even if they don't currently have DSL helpers), and what the text they should map to is.

From a quick skim (potentially non-exhaustive)

- `depends_on: "--foo"`
- `conflicts "--foo", "--bar"
- `env: :display_install_times`
- if the flag implies other flags (no 'constraint' DSL for this yet, maybe `implies "--foo", "--bar"`)
- if the flag is only for Formula, or only for Casks (no 'constraint' DSL for this yet, maybe `only_for: :cask`, `only_for: :formula`)
